src/libraries/Common/src/System/Net/Http/Http3/Helpers/VariableLengthIntegerHelper.cs

```diff
@@ -0,0 +1,213 @@
+            {
+                byte firstByte = buffer[0];
+
+                switch (firstByte & LengthMask)
+                {
+                    case LengthOneByte:
+                        value = firstByte;
+                        bytesRead = 1;
+                        return true;
+                    case LengthTwoByte:
+                        if (BinaryPrimitives.TryReadUInt16BigEndian(buffer, out ushort serializedShort))
+                        {
+                            value = serializedShort - TwoByteSubtract;
+                            bytesRead = 2;
+                            return true;
+                        }
+                        break;
+                    case LengthFourByte:
+                        if (BinaryPrimitives.TryReadUInt32BigEndian(buffer, out uint serializedInt))
+                        {
+                            value = serializedInt - FourByteSubtract;
+                            bytesRead = 4;
+                            return true;
+                        }
+                        break;
+                    default: // LengthEightByte
+                        Debug.Assert((firstByte & LengthMask) == LengthEightByte);
+                        if (BinaryPrimitives.TryReadUInt64BigEndian(buffer, out ulong serializedLong))
+                        {
+                            value = (long)(serializedLong - EightByteSubtract);
+                            Debug.Assert(value >= 0 && value <= EightByteLimit, "Serialized values are within [0, 2^62).");
+
+                            bytesRead = 8;
+                            return true;
+                        }
+                        break;
+                }
+            }
+
+            value = 0;
+            bytesRead = 0;
+            return false;
+        }
+
+        public static bool TryRead(ref SequenceReader<byte> reader, out long value)
+        {
+            // Hot path: we probably have the entire integer in one unbroken span.
+            if (TryRead(reader.UnreadSpan, out value, out int bytesRead))
+            {
+                reader.Advance(bytesRead);
+                return true;
+            }
+
+            // Cold path: copy to a temporary buffer before calling span-based read.
+            return TryReadSlow(ref reader, out value);
+
+            static bool TryReadSlow(ref SequenceReader<byte> reader, out long value)
+            {
+                ReadOnlySpan<byte> span = reader.CurrentSpan;
+
+                if (reader.TryPeek(out byte firstByte))
+                {
+                    int length =
+                        (firstByte & LengthMask) switch
+                        {
+                            LengthOneByte => 1,
+                            LengthTwoByte => 2,
+                            LengthFourByte => 4,
+                            _ => 8 // LengthEightByte
+                    };
+
+                    Span<byte> temp = (stackalloc byte[8])[..length];
```

Comment:
Why do we need to slice it?

Reply:
`(stackalloc byte[8])[..length]` has marginally better codegen than `stackalloc byte[length]`.

Reply:
I wasn't questioning that part, but rather why does it need to be perfectly sized at all.  Why not just `stackalloc byte[8]`?

Reply:
Oh. `SequenceReader.TryCopyTo` reads in the full buffer or fails, so 8 would cause EOF scenario to fail.

Reply:
I don't understand. `temp` here is the target of TryCopyTo, not the source.  What am I missing?

Reply:
It's about L121 with `TryRead(temp, out ...)`.

Althoug I have to admit that I don't understand why it is copied first, than read from. But that's a different story -- me missing some context here.

Reply:
> It's about L121 with TryRead(temp, out ...)

What about it?  That TryRead is taking a span, not a reader, and it's not using a reader internally.

Reply:
Ah, with the current implementation of `TryRead` it will work with any large enough size. So you're correct, slicing to actual size isn't needed.

(I believe to remember there were a version of TryRead expecting the actual size to work -- but irrelevant right now.)

Reply:
Slice is required here because of `reader.TryCopyTo`:

- Our stream has reach EOF and the reader exactly 2 bytes in it, containing a 2-byte integer.
- `reader.TryCopyTo` -- which succeeds only if it can fill the entire buffer -- fails, because there are not 8 bytes available.
- Upstream we see EOF but `VariableLengthIntegerHelper.TryRead` fails, so it throws something indicating an incomplete stream.

Reply:
> which succeeds only if it can fill the entire buffer

Ugh.  That is not an intuitive design.  That's not how any other "CopyTo" I'm aware of behaves.

Reply:
It got @GrabYourPitchforks too. Somehow nobody caught it during API review.